Try to improve remote bitbang speed again

timsifive opened the pull request:
For whatever reason my benchmarks all show this doesn't make a difference, and I can't even replicate the speed regression.

However, without a doubt the code now performs fewer system calls, so it must be faster, right?

timsifive committed:
Tweak remote bitbang code.

Reducing the number of times we perform two read(2) calls in a row.

Change-Id: If8f326d4f2c4ece979af427a2213f77b9024e6dd

timsifive committed:
Remove unnecessary block/non-block changes.

Change-Id: I095717d19483fb0daf2e6c264cab1134fd5f84cd
Signed-off-by: Tim Newsome <tim@sifive.com>

JanMatCodasip commented on src/jtag/drivers/remote_bitbang.c:
```diff
@@ -188,16 +189,14 @@ static int remote_bitbang_sample(void)
 static bb_value_t remote_bitbang_read_sample(void)
 {
 	if (remote_bitbang_recv_buf_start == remote_bitbang_recv_buf_end) {
```
This could be put to a separate function `remote_bitbang_recv_buf_is_empty()` for readability.

JanMatCodasip commented on src/jtag/drivers/remote_bitbang.c:
```diff
@@ -51,10 +51,48 @@ static bool remote_bitbang_buf_full(void)
+
+	bool first = true;
 	while (!remote_bitbang_buf_full()) {
 		unsigned int contiguous_available_space;
```
The computation of contiguous_available_space could also go to a separate function (for code readability).

JanMatCodasip commented on src/jtag/drivers/remote_bitbang.c:
```diff
@@ -188,16 +189,14 @@ static int remote_bitbang_sample(void)
+	assert(remote_bitbang_recv_buf_start != remote_bitbang_recv_buf_end);
+	int c = remote_bitbang_recv_buf[remote_bitbang_recv_buf_start];
+	remote_bitbang_recv_buf_start =
+		(remote_bitbang_recv_buf_start + 1) % sizeof(remote_bitbang_recv_buf);
```
Another snippet that should go to its own function - `remote_bitbang_recv_buf_pop_char()` or similar name.

JanMatCodasip reviewed (commented):
Checked visually, the code looks all right. I have just few suggestions towards better code readability.

JanMatCodasip commented:
One more idea that may potentially boost the performance: `TCP_NODELAY`. In my opinion, it is worth a try.

It is explained e.g. here: https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux_for_real_time/7/html/tuning_guide/tcp_nodelay_and_small_buffer_writes

See also: https://github.com/riscv/riscv-openocd/blob/riscv/src/jtag/drivers/jtag_vpi.c#L573

timsifive committed:
Refactor, creating remote_bitbang_recv_buf_empty()

Change-Id: Ib678bda990b3e224abe8faec2c48401b08cf4698

timsifive committed:
Refactor available space computation.

Creating remote_bitbang_recv_buf_contiguous_available_space() which has
a ridiculously long name but I don't see anything much better without
replacing the remote_bitbang prefix throughout the file.

Change-Id: I5d241b872a50a80a2df69fa60635207496484b7d

timsifive committed:
Use TCP_NODELAY on remote bitbang socket.

`make` in riscv-tests/debug went from 2m40s to 1m20s.

Change-Id: I0fa9f0007774626f2306d554397877c2117e8bdb
Signed-off-by: Tim Newsome <tim@sifive.com>

timsifive commented:
> One more idea that may potentially boost the performance: TCP_NODELAY. In my opinion, it is worth a try.

That was a huge improvement! Thanks!

timsifive committed:
Fix build on Windows.

Change-Id: I91f3ddee794e8540c5e871ead3c09583c272bd34
Signed-off-by: Tim Newsome <tim@sifive.com>

GregSavin reviewed (approved):
I wonder if the slowdown that was reported was specific to a particular operating system.  In any case, that was a good idea to try TCP_NODELAY.
